crates/bevy_app/src/task_pool_options.rs

```diff
@@ -0,0 +1,157 @@
+        // available, this is intended. An example case where this might happen is a device with
+        // <= 2 threads.
+        clamp_usize(desired, self.min_threads, self.max_threads)
+    }
+}
+
+/// Helper for configuring and creating the default task pools. For end-users who want full control,
+/// insert the default task pools into the resource map manually. If the pools are already inserted,
+/// this helper will do nothing.
+#[derive(Clone)]
+pub struct DefaultTaskPoolOptions {
+    /// If the number of physical cores is less than min_total_threads, force using min_total_threads
+    pub min_total_threads: usize,
+    /// If the number of physical cores is grater than max_total_threads, force using max_total_threads
+    pub max_total_threads: usize,
+
+    /// Used to determine number of IO threads to allocate
+    pub io: TaskPoolThreadAssignmentPolicy,
+    /// Used to determine number of async compute threads to allocate
+    pub async_compute: TaskPoolThreadAssignmentPolicy,
+    /// Used to determine number of compute threads to allocate
+    pub compute: TaskPoolThreadAssignmentPolicy,
+}
+
+impl Default for DefaultTaskPoolOptions {
+    fn default() -> Self {
+        DefaultTaskPoolOptions {
+            // By default, use however many cores are available on the system
+            min_total_threads: 1,
+            max_total_threads: std::usize::MAX,
+
+            // Use 25% of cores for IO, at least 1, no more than 4
+            io: TaskPoolThreadAssignmentPolicy {
+                min_threads: 1,
+                max_threads: 4,
+                percent: 0.25,
+            },
+
+            // Use 25% of cores for async compute, at least 1, no more than 4
+            async_compute: TaskPoolThreadAssignmentPolicy {
+                min_threads: 1,
+                max_threads: 4,
+                percent: 0.25,
+            },
+
+            // Use all remaining cores for compute (at least 1)
+            compute: TaskPoolThreadAssignmentPolicy {
+                min_threads: 1,
+                max_threads: std::usize::MAX,
+                percent: 1.0, // This 1.0 here means "whatever is left over"
+            },
+        }
+    }
+}
+
+impl DefaultTaskPoolOptions {
+    /// Create a configuration that forces using the given number of threads.
+    pub fn with_num_threads(thread_count: usize) -> Self {
+        let mut options = Self::default();
+        options.min_total_threads = thread_count;
+        options.max_total_threads = thread_count;
+
+        options
+    }
+
+    /// Inserts the default thread pools into the given resource map based on the configured values
+    pub fn create_default_pools(&self, resources: &mut Resources) {
+        let total_threads = clamp_usize(
+            num_cpus::get(),
+            self.min_total_threads,
+            self.max_total_threads,
+        );
+
+        let mut remaining_threads = total_threads;
+
+        if !resources.contains::<IOTaskPool>() {
```

Comment:
Ordering determining "precedence" is maybe a little hard to reason about here (given that this is abstracted out from the user). what if someone configures 100% for io threads? it seems like that would prevent AsyncComputeTaskPool and ComputeTaskPool from getting anything more than the min thread count of `1`, even if they are also 100%. 

I'm not sure what the "fix" would be here, other than some sort of "global" view of pools that would use percentages as weights. But that introduces additional complexity that I'm pretty sure isn't worth it.

Another alternative is to just completely remove "remaining_threads" and instead let users configure pools in a way that goes over the total number of available cores.

Reply:
Yeah this is tricky. @lachlansneff and I bounced a few ideas back and forth:

* Do something stupid-simple like 1 IO, 1 async-compute, and N-2 compute
* Something arguably slightly less stupid but still simple (implementation that's in the PR)
* Force the end user to deal with it with some kind of callback

The main reason for this "precedence" approach is that it avoids an awkward situation where due to fractions/core count it becomes difficult to pick ratios that for every hardware configuration produce 1:1 ration of cores and threads.

End-users can always push their own pools into the resources so I think that takes some pressure off of us having an implementation that can do everything for everyone. I would favor towards keeping things simple. I also think it's reasonable to expect anyone tweaking something like this to read the code.

Reply:
Ok thats reasonable. The default behavior is simple enough (if you read the code) and overriding is possible. Lets leave it as-is for now.